### DISCUSSION TITLE
What is best way to deal with client network errors?

### DISCUSSION THREAD

**Discussion by Olex1313** [bluenviron/gortsplib#239] [Questions] [answered]
Hello, I'm currently working on my course project in university and my goal is to write simple rtsp proxy with error-prone client. 

Generally speaking I have a ip-cam with rtsp stream in lecture room and ffmpeg-based scheduled scripts somewhere to save lecture record. But due to network issues sometimes camera is unavailable which causes ffmpeg to throw an error.

I want to put a proxy between camera and client, and when network error occurs send some picture with "please stand by". I'm not that familiar with network programming and RTSP protocol, but can I just catch error after play request and directly write packets to ServerStream? Or is there some way to do it more clearly?
```
// start playing
	_, err = rc.Play(nil)
	if err.Error() == "EOF" {
		stream.WritePacketRTP()
	}
	if err != nil {
		return err // "EOF"
	}
```

**Comment by aler9** [comment, marked as answer]
Hello, this task is difficult since it involves creating a continuous and coherent H264 stream (assuming your camera uses H264 as its video codec). The RTSP protocol only implements the outer layer that wraps the H264 stream. When a camera disconnects, the H264 stream goes missing and it's impossible to recreate it or insert frames in between.

The easiest way i can think of implementing this consists in:

1. Connect to the stream, start decoding H264 access units:
    https://github.com/bluenviron/gortsplib/blob/main/examples/client-read-format-h264/main.go

2. Create a server that encodes the H264 access units and makes them available to any client:
    https://github.com/bluenviron/gortsplib/blob/main/examples/server/main.go

    H264 encoder is here:
    https://pkg.go.dev/github.com/bluenviron/gortsplib/v3@v3.1.0/pkg/formats/rtph264#Encoder

2. Keep the last H264 access unit stored in memory

4. When the RTSP client disconnects, start a routine that periodically publishes the last H264 access unit, and in the meanwhile start another RTSP client

5. When the new RTSP client is connected and ready, stop the routine and attach the new RTSP client to the server again

**Reply by Olex1313** [reply]
Thanks, I'll try to make some cache for packages and will try to re-send it

**Reply by danielamar101** [reply]
Were you able to implement this approach @Olex1313 ?

**Reply by derSoerrn95** [reply]
@Olex1313 did you solve it?

**Reply by Olex1313** [reply]
@derSoerrn95, @danielamar101, Sorry for late reply. It's been a long time since my university struggles with rtsp :), the implicit caching of RTP packets and resending it is somewhat implemented in https://github.com/Olex1313/strmr, 

However the approach itself isn't quite right, because when the proxied source re-establishes connection with your proxy -> the SSRC's for media sources in the session will change and you need to explicitly handle that.

Overall, the "unreliable client" problem is better to be solved with realtime muxing server, which takes in account the inner content of packets, to maintain SSRC's and timestamps for given codec, the "raw-RTP" approach is incorrect and mostly will not work, due to different codec profiles, change of codecs itself, etc. 

If you are trying to build a reliable proxy - I would suggest you to take libav if you are aiming for wide codec support, or gortsplib with h264 codec implementation and maintain two "fake" session to upstream (publishes to proxy) and downstream (reads from proxy), and then decode the packets, muxing them, and sending the last received with maintaining the timestamps offset.

If you are trying to build a reliable recorder - the best approach is to record separate chunks and then merge them with ffmpeg

**Reply by derSoerrn95** [reply]
Hi @Olex1313,

thanks for your detailed answer. I'm going to look into it. But maybe I try GStreamer first.
